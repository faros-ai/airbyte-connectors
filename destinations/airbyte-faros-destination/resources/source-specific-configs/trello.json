{
  "title": "Trello",
  "type": "object",
  "oneOf": [
    {
      "title": "Configuration",
      "type": "object",
      "properties": {
        "source_type": {
          "type": "string",
          "const": "Trello",
          "order": 0
        },
        "task_status_category_mapping": {
          "type": "string",
          "title": "Task Status Category Mapping",
          "description": "JSON map from Trello list to Faros task status category. The keys are the Trello list names or ids and the values are the Faros task status categories.",
          "multiline": true,
          "default": "{}",
          "examples": [
            "{\"New\": \"Todo\", \"Ongoing\": \"InProgress\", \"Completed\": \"Done\"}",
            "{\"108e25a66845e60ab267f554\": \"Todo\", \"57cfc2f7873dcd4aeab2ae15\": \"InProgress\", \"eb6ab183fe470ff75cb1c902\": \"Done\"}"
          ],
          "order": 1
        }
      }
    }
  ]
}
