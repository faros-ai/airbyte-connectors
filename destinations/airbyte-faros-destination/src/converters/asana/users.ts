import {AirbyteRecord} from 'faros-airbyte-cdk';

import {DestinationModel, DestinationRecord} from '../converter';
import {AsanaCommon, AsanaConverter, AsanaUser} from './common';

export class Users extends AsanaConverter {
  readonly destinationModels: ReadonlyArray<DestinationModel> = ['tms_User'];

  async convert(
    record: AirbyteRecord
  ): Promise<ReadonlyArray<DestinationRecord>> {
    const source = this.streamName.source;
    const user = record.record.data as AsanaUser;

    return [AsanaCommon.tms_User(user, source)];
  }
}
