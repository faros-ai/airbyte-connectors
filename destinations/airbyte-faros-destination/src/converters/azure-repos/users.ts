import {AirbyteRecord} from 'faros-airbyte-cdk';

import {DestinationModel, DestinationRecord} from '../converter';
import {AzureReposConverter} from './common';
import {User, UserType, UserTypeCategory} from './models';

export class Users extends AzureReposConverter {
  readonly destinationModels: ReadonlyArray<DestinationModel> = [
    'vcs_Membership',
    'vcs_User',
  ];

  async convert(
    record: AirbyteRecord
  ): Promise<ReadonlyArray<DestinationRecord>> {
    const source = this.streamName.source;
    const userItem = record.record.data as User;
    const res: DestinationRecord[] = [];
    const organizationName = this.getOrganizationFromUrl(
      userItem._links.self.href
    );
    const organization = {uid: organizationName, source};
    const type: UserType = {
      category: UserTypeCategory.User,
      detail: userItem.subjectKind,
    };
    res.push({
      model: 'vcs_Membership',
      record: {
        organization,
        user: {uid: userItem.principalName, source},
      },
    });

    res.push({
      model: 'vcs_User',
      record: {
        uid: userItem.principalName,
        name: userItem.displayName,
        type,
        htmlUrl: userItem.url,
        source,
      },
    });
    return res;
  }
}
