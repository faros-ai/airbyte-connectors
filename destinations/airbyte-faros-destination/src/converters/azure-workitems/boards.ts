import {AirbyteRecord} from 'faros-airbyte-cdk';

import {DestinationModel, DestinationRecord} from '../converter';
import {AzureWorkitemsConverter} from './common';
import {Board} from './models';

export class Boards extends AzureWorkitemsConverter {
  readonly destinationModels: ReadonlyArray<DestinationModel> = [
    'tms_TaskBoard',
  ];

  async convert(
    record: AirbyteRecord
  ): Promise<ReadonlyArray<DestinationRecord>> {
    const Board = record.record.data as Board;
    return [
      {
        model: 'tms_TaskBoard',
        record: {
          uid: String(Board.id),
          name: Board.name,
        },
      },
    ];
  }
}
