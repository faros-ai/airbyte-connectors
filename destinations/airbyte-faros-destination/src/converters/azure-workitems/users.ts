import {AirbyteRecord} from 'faros-airbyte-cdk';

import {DestinationModel, DestinationRecord} from '../converter';
import {AzureWorkitemsConverter} from './common';
import {User} from './models';

export class Users extends AzureWorkitemsConverter {
  readonly destinationModels: ReadonlyArray<DestinationModel> = ['tms_User'];

  async convert(
    record: AirbyteRecord
  ): Promise<ReadonlyArray<DestinationRecord>> {
    const source = this.streamName.source;
    const userItem = record.record.data as User;
    const res: DestinationRecord[] = [];

    res.push({
      model: 'tms_User',
      record: {
        uid: userItem.principalName,
        name: userItem.displayName,
        emailAddress: userItem.mailAddress,
        source,
      },
    });
    return res;
  }
}
