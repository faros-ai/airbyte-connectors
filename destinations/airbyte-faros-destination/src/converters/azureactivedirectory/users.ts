import {AirbyteRecord} from 'faros-airbyte-cdk';
import {Utils} from 'faros-js-client';

import {DestinationModel, DestinationRecord} from '../converter';
import {AzureActiveDirectoryConverter} from './common';
import {User} from './models';

export class Users extends AzureActiveDirectoryConverter {
  readonly destinationModels: ReadonlyArray<DestinationModel> = [
    'geo_Address',
    'geo_Location',
    'identity_Identity',
    'org_Department',
    'org_Employee',
  ];

  private seenDepartments = new Set<string>();

  async convert(
    record: AirbyteRecord
  ): Promise<ReadonlyArray<DestinationRecord>> {
    const source = this.streamName.source;
    const user = record.record.data as User;
    const joinedAt = Utils.toDate(user.createdDateTime);
    const manager = user.manager ? {uid: user.manager, source} : undefined;
    const uid = user.id;
    const res: DestinationRecord[] = [];

    if (user.department && !this.seenDepartments.has(user.department)) {
      this.seenDepartments.add(user.department);
      res.push({
        model: 'org_Department',
        record: {
          uid: user.department,
          name: user.department,
        },
      });
    }

    res.push({
      model: 'identity_Identity',
      record: {
        uid,
        fullName: `${user.givenName} ${user.surname}`,
        lastName: user.surname,
        primaryEmail: user.mail,
        emails: [user.mail],
      },
    });

    const location = user.streetAddress
      ? {uid: user.streetAddress, source}
      : undefined;

    if (user.streetAddress) {
      res.push({
        model: 'geo_Address',
        record: {
          uid: user.streetAddress,
          fullAddress: user.streetAddress,
          street: user.streetAddress,
          postalCode: user.postalCode,
        },
      });
      const geo_Location = {
        uid: user.streetAddress,
        name: user.streetAddress,
        raw: user.streetAddress,
        address: {uid: user.streetAddress},
      };
      res.push({model: 'geo_Location', record: geo_Location});
    }
    res.push({
      model: 'org_Employee',
      record: {
        uid,
        title: user.displayName,
        level: 0,
        joinedAt,
        department: user.department ? {uid: user.department} : null,
        identity: {uid, source},
        manager,
        location,
        source,
      },
    });
    return res;
  }
}
