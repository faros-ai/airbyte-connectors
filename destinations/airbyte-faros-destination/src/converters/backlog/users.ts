import {AirbyteRecord} from 'faros-airbyte-cdk';

import {DestinationModel, DestinationRecord} from '../converter';
import {BacklogConverter} from './common';
import {User} from './models';

export class Users extends BacklogConverter {
  readonly destinationModels: ReadonlyArray<DestinationModel> = ['tms_User'];

  async convert(
    record: AirbyteRecord
  ): Promise<ReadonlyArray<DestinationRecord>> {
    const source = this.streamName.source;
    const user = record.record.data as User;
    return [
      {
        model: 'tms_User',
        record: {
          uid: String(user.id),
          name: user.name,
          emailAddress: user.mailAddress,
          source,
        },
      },
    ];
  }
}
