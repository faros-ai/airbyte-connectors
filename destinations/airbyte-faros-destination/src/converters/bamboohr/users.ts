import {AirbyteRecord} from 'faros-airbyte-cdk';
import {Utils} from 'faros-js-client';
import {intersection, uniq} from 'lodash';

import {DestinationModel, DestinationRecord, StreamContext} from '../converter';
import {BambooHRConverter} from './common';
import {User} from './models';

const ROOT_TEAM_UID = 'all_teams';

export class Users extends BambooHRConverter {
  readonly destinationModels: ReadonlyArray<DestinationModel> = [
    'geo_Address',
    'geo_Location',
    'identity_Identity',
    'org_Department',
    'org_Employee',
    'org_Team',
    'org_TeamMembership',
  ];

  private seenDepartments = new Set<string>();
  private managers = new Map<string, string>();
  private employeeNamesById = new Map<string, string>();

  async convert(
    record: AirbyteRecord,
    ctx: StreamContext
  ): Promise<ReadonlyArray<DestinationRecord>> {
    const inactiveEmploymentHistoryStatus =
      this.inactiveEmploymentHistoryStatus(ctx);
    const source = this.streamName.source;
    const user = record.record.data as User;
    const uid = user.id;
    let joinedAt = Utils.toDate(user.hireDate);
    if (isNaN(joinedAt?.getTime())) {
      ctx.logger.warn(
        `Found unexpected hire date ${user.hireDate} for user id ${user.id}`
      );
      joinedAt = null;
    }
    let terminatedAt: Date;
    if (user.terminationDate === '0000-00-00') {
      terminatedAt = null;
    } else {
      terminatedAt = Utils.toDate(user.terminationDate);
      if (isNaN(terminatedAt?.getTime())) {
        ctx.logger.warn(
          `Found unexpected termination date ${user.terminationDate} for user id ${user.id}`
        );
        terminatedAt = null;
      }
    }
    const manager = user.supervisorEId ? {uid: user.supervisorEId} : undefined;
    const res: DestinationRecord[] = [];

    if (this.bootstrapTeamsFromManagers(ctx)) {
      this.employeeNamesById.set(uid, user.fullName1);
      if (manager) {
        this.managers.set(uid, manager.uid);
      }
    }

    if (user.department && !this.seenDepartments.has(user.department)) {
      this.seenDepartments.add(user.department);
      res.push({
        model: 'org_Department',
        record: {
          uid: user.department,
          name: user.department,
        },
      });
    }

    res.push({
      model: 'identity_Identity',
      record: {
        uid,
        firstName: user.firstName,
        fullName: user.fullName1,
        lastName: user.lastName,
        primaryEmail: user.bestEmail ?? user.workEmail,
        emails: uniq([user.bestEmail, user.workEmail].filter((e) => e)),
      },
    });

    const fullAddress = [user.city, user.stateCode ?? user.state, user.country]
      .filter((a) => a)
      .join(', ');
    const location = fullAddress ? {uid: fullAddress} : undefined;

    if (fullAddress) {
      res.push({
        model: 'geo_Address',
        record: {
          uid: fullAddress,
          fullAddress,
          city: user.city,
          state: user.state,
          stateCode: user.stateCode,
          country: user.country,
        },
      });
      const geo_Location = {
        uid: fullAddress,
        name: fullAddress,
        raw: fullAddress,
        address: {uid: fullAddress},
      };
      res.push({model: 'geo_Location', record: geo_Location});
    }

    let inactive = false;
    if (inactiveEmploymentHistoryStatus?.length) {
      inactive = inactiveEmploymentHistoryStatus.includes(
        user.employmentHistoryStatus
      );
    } else if (user.status) {
      inactive = user.status.toLowerCase() === 'inactive';
    }

    res.push({
      model: 'org_Employee',
      record: {
        uid,
        title: user.jobTitle,
        joinedAt,
        department: user.department ? {uid: user.department} : null,
        identity: {uid, source},
        inactive,
        manager,
        location,
        terminatedAt,
        source,
      },
    });
    return res;
  }

  async onProcessingComplete(
    ctx: StreamContext
  ): Promise<ReadonlyArray<DestinationRecord>> {
    const res: DestinationRecord[] = [];
    if (!this.bootstrapTeamsFromManagers(ctx)) return res;

    for (const uid of intersection(
      Array.from(this.managers.values()),
      Array.from(this.employeeNamesById.keys())
    )) {
      let parentTeamId = this.managers.get(uid);
      if (!parentTeamId || !this.employeeNamesById.has(parentTeamId)) {
        parentTeamId = ROOT_TEAM_UID;
      }
      res.push({
        model: 'org_Team',
        record: {
          uid,
          name: `${this.employeeNamesById.get(uid)} Org`,
          lead: {uid},
          parentTeam: {uid: parentTeamId},
        },
      });
      res.push({
        model: 'org_TeamMembership',
        record: {
          team: {uid},
          member: {uid},
        },
      });
    }

    for (const uid of this.employeeNamesById.keys()) {
      const managerUid = this.managers.get(uid);
      if (managerUid && this.employeeNamesById.has(managerUid)) {
        res.push({
          model: 'org_TeamMembership',
          record: {
            team: {uid: managerUid},
            member: {uid},
          },
        });
      }
    }

    return res;
  }
}
