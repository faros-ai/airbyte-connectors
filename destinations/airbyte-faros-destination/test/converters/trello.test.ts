import {AirbyteRecord} from 'faros-airbyte-cdk';

import {Boards} from '../../src/converters/trello/boards';
import {Users} from '../../src/converters/trello/users';
import {generateBasicTestSuite} from './utils';

generateBasicTestSuite({sourceName: 'trello'});

describe('trello', () => {
  describe('boards', () => {
    const converter = new Boards();
    const BOARD = {
      id: 'b98e01d14cf778006ae8e8d1',
      name: 'Innovative Venture',
    };

    test('basic board', async () => {
      const record = AirbyteRecord.make('boards', BOARD);
      const res = await converter.convert(record);
      expect(res).toMatchSnapshot();
    });
  });

  describe('users', () => {
    const converter = new Users();
    const USER = {
      id: '1735be8ade395c72d43d6603',
      fullName: 'Jamie Williams',
      username: 'jwilliams80',
    };

    test('basic user', async () => {
      const record = AirbyteRecord.make('mytestsource__trello__users', USER);
      const res = await converter.convert(record);
      expect(res).toMatchSnapshot();
    });
  });
});
