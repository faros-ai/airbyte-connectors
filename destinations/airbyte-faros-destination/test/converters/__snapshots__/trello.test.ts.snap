// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`trello basic test process records from all streams 1`] = `
Array [
  "Processed 2 records",
  "Processed records by stream: {\\\\\\"mytestsource__trello__boards\\\\\\":1,\\\\\\"mytestsource__trello__users\\\\\\":1}\\"},\\"type\\":\\"LOG\\"}",
  "Would write 4 records",
  "Would write records by model: {\\\\\\"tms_Project\\\\\\":1,\\\\\\"tms_TaskBoard\\\\\\":1,\\\\\\"tms_TaskBoardProjectRelationship\\\\\\":1,\\\\\\"tms_User\\\\\\":1}\\"},\\"type\\":\\"LOG\\"}",
  "Skipped 0 records",
  "Errored 0 records",
]
`;

exports[`trello boards basic board 1`] = `
Array [
  Object {
    "model": "tms_TaskBoard",
    "record": Object {
      "name": "Innovative Venture",
      "source": "Trello",
      "uid": "b98e01d14cf778006ae8e8d1",
    },
  },
  Object {
    "model": "tms_Project",
    "record": Object {
      "name": "Innovative Venture",
      "source": "Trello",
      "uid": "b98e01d14cf778006ae8e8d1",
    },
  },
  Object {
    "model": "tms_TaskBoardProjectRelationship",
    "record": Object {
      "board": Object {
        "source": "Trello",
        "uid": "b98e01d14cf778006ae8e8d1",
      },
      "project": Object {
        "source": "Trello",
        "uid": "b98e01d14cf778006ae8e8d1",
      },
    },
  },
]
`;

exports[`trello users basic user 1`] = `
Array [
  Object {
    "model": "tms_User",
    "record": Object {
      "emailAddress": "jwilliams80",
      "name": "Jamie Williams",
      "source": "Trello",
      "uid": "1735be8ade395c72d43d6603",
    },
  },
]
`;
